content/backend/typescript-apollo/7-voting-and-scalars.md

```diff
@@ -0,0 +1,524 @@
+            resolve(parent, args, context) {
+                return context.prisma.user
+                    .findUnique({ where: { id: parent.id } })
+                    .votes();
+            }
+        })
+    },
+});
+
+```
+
+
+</Instruction>
+
+All right, that’s it! Now let's test the implementation of the `vote` mutation! When executing the following instruction, make sure the `Authorization` header is enabled and has a valid token. 
+
+
+<Instruction>
+
+Go to Apollo Studio Explorer and execute the following mutation:
+
+```graphql
+mutation {
+  vote(linkId: _LINK_ID_) {
+    link {
+      url
+      description
+    }
+    user {
+      name
+      email
+    }
+  }
+}
+```
+
+</Instruction>
+
+You should see an output similar to the following: 
+
+```json(nocopy)
+{
+  "data": {
+    "vote": {
+      "link": {
+        "url": "nexusjs.org",
+        "description": "Code-First GraphQL schemas for JavaScript/TypeScript"
+      },
+      "user": {
+        "name": "alice",
+        "email": "alice@prisma.io"
+      }
+    }
+  }
+}
+```
+
+![Vote mutation](https://i.imgur.com/oIIXH9C.png)
+
+
+Awesome! Now it's time to learn about GraphQL scalars. 
+
+
+### What are GraphQL scalars? 
+
+As mentioned before, scalars are the basic types in a GraphQL schema, similar to the primitive types in programming languages. While doing this tutorial, you have used a few of the built-in scalar types, notably `String` and `Int`. 
+
+While the built-in scalars support most of the common use-cases, your application might need support for other custom scalars. For example, currently in our application, the `Link.url` field has a `String` scalar type. However, you might want to create an `Url` scalar to add custom validation logic and reject invalid urls. The benefit of scalars is that they simultaneously define the _representation_ and _validation_ for the primitive data types in your API. 
+
+
+
+### Custom scalars with Nexus
+
+Nexus allows you to create custom scalars for your GraphQL schema using a standardized interface. To do this, you will need to provide the functions that _serialize_ and _deserialize_ data for that scalar. It's important to note that under the hood Nexus uses the `GraphQLScalarType` class from the `graphql` library. Most applications in the node ecosystem uses this classs when defining custom scalars. As a result scalars defined across different libraries in the node ecosystem are often compatible with each other. 
+
+The [graphql-scalars](https://github.com/Urigo/graphql-scalars) library provides a large set of custom Scalars for creating precise type-safe GraphQL schemas. You are going to import and use the `DateTime` scalar from this library in your GraphQL API. 
+
+### Adding the `DateTime` scalar
+
+Start by adding the `graphql-scalars` library to your application. 
+
+<Instruction>
+
+Install graphql-scalars using npm: 
+
+```bash(path=".../hackernews-typescript/")
+npm install graphql-scalars^1.14.1
```

Comment:
Same issue as in my earlier review that the `@` seems to be needed for this command, otherwise this error is thrown for me:

```
$ npm install graphql-scalars^1.14.1
npm ERR! code EINVALIDTAGNAME
npm ERR! Invalid tag name "graphql-scalars^1.14.1": Tags may not have any characters that encodeURIComponent encodes.

```

```suggestion
npm install graphql-scalars@^1.14.1
```